quinn/src/platform/unix.rs

```diff
@@ -198,9 +198,18 @@ fn send(io: &mio::net::UdpSocket, transmits: &[Transmit]) -> io::Result<usize> {
     let mut msgs: [libc::mmsghdr; BATCH_SIZE] = unsafe { mem::zeroed() };
     let mut iovecs: [libc::iovec; BATCH_SIZE] = unsafe { mem::zeroed() };
     let mut cmsgs = [cmsg::Aligned([0u8; CMSG_LEN]); BATCH_SIZE];
+    let mut addrs: [MaybeUninit<socket2::SockAddr>; BATCH_SIZE] = unsafe {
+        MaybeUninit::uninit().assume_init()
```

Comment:
This defeats the purpose of `MaybeUninit`. You should not call `assume_init` until the value has been properly initialized. Which you do further down. Also I think type inference plays a trick here and the value is not what you expect it to be unless I'm misunderstanding. The type of `addrs` is supposed to be uninitialized, but you assume init it. You probably want:
```rust
let mut addrs: [MaybeUninit<socket2::SockAddr>; BATCH_SIZE] = [MaybeUninit::uninit(); BATCH_SIZE];
```

Reply:
I took it from the [MaybeUninit docs](https://doc.rust-lang.org/stable/core/mem/union.MaybeUninit.html#initializing-an-array-element-by-element). Basically, the `MaybeUninit::uninit().assume_init()` desugars to `MaybeUninit::<[MaybeUninit<socket2::SockAddr>; BATCH_SIZE]>::uninit().assume_init()`. Doing that it safe according to the docs because the inner MaybeUninit is allowed to be uninitialized.

Your suggestion doesn't compile on stable because SockAddr does not impl Copy and the const expr repeat RFC is not stable (and even if it were, I'd like something that compiles on older MSRVs too as I want to backport this patch).

Yes, the MaybeUninit APIs around arrays are a bit weird. There are a couple of unstable proposals to improve upon this like [`uninit_array`](https://doc.rust-lang.org/stable/core/mem/union.MaybeUninit.html#method.uninit_array) but we can't use those due to their unstable-ness :).

Reply:
Oh darn. Did not realize `MaybeUninit<T>` was not always `Copy`. You are right. But very very strange still. We really need const repeat for sure.